### PR TITLE
Migrate from Flake8 & Autopep8 to Ruff

### PR DESCRIPTION
**Pull request by vdusek**
### Description

- Ruff presents an extensive set of rules. I decided to adopt a strategy of enabling all rules by default and selectively disabling specific rules or entire rule groups as needed (either for the whole project/directory/specific file). 
- The code was automatically formatted using `ruff format` and `ruff check --fix`, some easy fixes were performed manually, and some rule violations were turned off for the particular line of code.
    - Including the usage of `from __future__ import annotations` which will be discussed separately.
- Also `Makefile`, `CONTRIBUTING.md`, `.gitignore`, and a few other files were slightly updated.

### Issue

- Closes https://github.com/apify/apify-shared-python/issues/15

### CODE REVIEW

File: src/apify_shared/models.py [hunk 1]
```diff
@@ -1,4 +1,6 @@
-from typing import Dict, Generic, List, TypeVar
+from __future__ import annotations
```
Comment (jirimoravcik) [inline src/apify_shared/models.py:1]: I like this and I'm not against it. Not sure if something could go wrong with this. But if lint and tests pass, I guess it's fine.

File: tests/unit/test_utils.py [hunk 2]
```diff
@@ -31,17 +31,18 @@ class TestEnum(Enum):
...
 
 
-def test__filter_out_none_values_recursively_internal() -> None:  # Copypasted from client
-    assert _filter_out_none_values_recursively_internal({}) == {}
-    assert _filter_out_none_values_recursively_internal({'k1': {}}) == {}
-    assert _filter_out_none_values_recursively_internal({}, False) == {}
-    assert _filter_out_none_values_recursively_internal({'k1': {}}, False) == {'k1': {}}
-    assert _filter_out_none_values_recursively_internal({}, True) is None
-    assert _filter_out_none_values_recursively_internal({'k1': {}}, True) is None
+def test_filter_out_none_values_recursively_internal() -> None:  # Copypasted from client
```
Comment (jirimoravcik) [inline tests/unit/test_utils.py:39]: Hmm, I'm thinking this lovely comment was created before we had the shared package, can you check if it's still correct to have something "copypasted"?

**Review by jirimoravcik — CHANGES_REQUESTED** [review]
Nice work, had some comments

File: Makefile [hunk 3]
```diff
@@ -18,22 +20,22 @@ twine-check:
 	python -m twine check dist/*
 
 lint:
-	python3 -m flake8
+	python -m ruff check $(DIRS_WITH_CODE)
```
[the remark this answers is not in the document]
Comment (vdusek) [reply Makefile:23]: Yeah... I saw there is already some usage of just `python` in the Makefile, so I wanted to ensure consistency and align everything. However, got it, gonna change everything to `python3` instead.

File: src/apify_shared/models.py (hunk 1 above)
Comment (vdusek) [reply src/apify_shared/models.py:1]: Yeah... linting and tests have passed, I personally tested it on Python 3.8, and it's also utilized in larger projects like Airflow (e.g. here [airflow/providers/sftp/operators/sftp.py#L19](https://github.com/apache/airflow/blob/0a93e2e28baa282e20e2a68dcb718e3708048a47/airflow/providers/sftp/operators/sftp.py#L19)), which supports Python 3.8 and follows a similar approach.

So hopefully we should be good :crossed_fingers:.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -24,11 +24,11 @@ To install this package and its development dependencies, run `make install-dev`
 
 ## Formatting
 
-We use `autopep8` and `isort` to automatically format the code to a common format. To run the formatting, just run `make format`.
+We use `ruff` and `isort` to automatically format the code to a common format. To run the formatting, just run `make format`.
```
[the remark this answers is not in the document]
Comment (vdusek) [reply CONTRIBUTING.md]: It seems we don't, `ruff` should replace it as well. So I removed the `isort` config as well, and everything works as before. Nice.

**Review by jirimoravcik — APPROVED** [review]

**Pull request merged**